operator/pkg/translate/translate.go

```diff
@@ -243,13 +247,157 @@ func (t *Translator) OverlayK8sSettings(yml string, iop *v1alpha1.IstioOperatorS
 		if err != nil {
 			return "", err
 		}
+
+		// Apply the workaround for merging service ports with (port,protocol) composite
+		// keys instead of just the merging by port.
+		if inPathParts[len(inPathParts)-1] == "Service" {
+			if msvc, ok := m.(*v1alpha1.ServiceSpec); ok {
+				mergedObj, err = t.fixMergedObjectWithCustomServicePortOverlay(oo, msvc, mergedObj)
+				if err != nil {
+					return "", err
+				}
+			}
+		}
+
 		// Update the original object in objects slice, since the output should be ordered.
 		*(om[pe]) = *mergedObj
 	}
 
 	return objects.YAMLManifest()
 }
 
+func (t *Translator) fixMergedObjectWithCustomServicePortOverlay(oo *object.K8sObject,
+	msvc *v1alpha1.ServiceSpec, mergedObj *object.K8sObject) (*object.K8sObject, error) {
+	var basePorts []*v1.ServicePort
+	bps, _, err := unstructured.NestedSlice(oo.Unstructured(), "spec", "ports")
+	if err != nil {
+		return nil, err
+	}
+	bby, err := json.Marshal(bps)
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(bby, &basePorts); err != nil {
+		return nil, err
+	}
+	overlayPorts := make([]*v1.ServicePort, 0, len(msvc.GetPorts()))
+	for _, p := range msvc.GetPorts() {
+		var pr v1.Protocol
+		switch strings.ToLower(p.GetProtocol()) {
+		case "udp":
+			pr = v1.ProtocolUDP
+		default:
+			pr = v1.ProtocolTCP
+		}
+		port := &v1.ServicePort{
+			Name:     p.GetName(),
+			Protocol: pr,
+			Port:     p.GetPort(),
+			NodePort: p.GetNodePort(),
+		}
+		if p.TargetPort != nil {
+			port.TargetPort = p.TargetPort.IntOrString
+		}
+		overlayPorts = append(overlayPorts, port)
+	}
+	mergedPorts := strategicMergePorts(basePorts, overlayPorts)
+	mpby, err := json.Marshal(mergedPorts)
+	if err != nil {
+		return nil, err
+	}
+	var mergedPortSlice []interface{}
+	if err = json.Unmarshal(mpby, &mergedPortSlice); err != nil {
+		return nil, err
+	}
+	if err = unstructured.SetNestedSlice(mergedObj.Unstructured(), mergedPortSlice, "spec", "ports"); err != nil {
+		return nil, err
+	}
+	// Now fix the merged object
+	mjsonby, err := json.Marshal(mergedObj.Unstructured())
+	if err != nil {
+		return nil, err
+	}
+	if mergedObj, err = object.ParseJSONToK8sObject(mjsonby); err != nil {
+		return nil, err
+	}
+	return mergedObj, nil
+}
+
+type portWithProtocol struct {
+	port     int32
+	protocol v1.Protocol
+}
+
+// strategicMergePorts merges the base with the given overlay considering both
+// port and the protocol as the merge keys. This is a workaround for the strategic
+// merge patch in Kubernetes which only uses port number as the key. This causes
+// an issue when we have to expose the same port with different protocols.
+// See - https://github.com/kubernetes/kubernetes/issues/103544
+func strategicMergePorts(base, overlay []*v1.ServicePort) []*v1.ServicePort {
```

Comment:
Instead of special hack, can we just sort it based on original order? Something like
```go
func indexOf(element portWithProtocol, data []portWithProtocol) int {
	for k, v := range data {
		if element == v {
			return k
		}
	}
	return len(data)
}

// strategicMergePorts merges the base with the given overlay considering both
// port and the protocol as the merge keys. This is a workaround for the strategic
// merge patch in Kubernetes which only uses port number as the key. This causes
// an issue when we have to expose the same port with different protocols.
// See - https://github.com/kubernetes/kubernetes/issues/103544
func strategicMergePorts(base, overlay []*v1.ServicePort) []*v1.ServicePort {
	portPriority := []portWithProtocol{}
	for _, p := range base {
		portPriority = append(portPriority, portWithProtocol{p.Port, p.Protocol})
	}
	for _, p := range overlay {
		portPriority = append(portPriority, portWithProtocol{p.Port, p.Protocol})
	}
	sortFn := func(ps []*v1.ServicePort) func(int, int) bool {
		return func(i, j int) bool {
			ii := indexOf(portWithProtocol{ps[i].Port, ps[i].Protocol}, portPriority)
			ji := indexOf(portWithProtocol{ps[i].Port, ps[i].Protocol}, portPriority)
			return ii < ji
		}
	}
```

I am just slightly worried that we will break users, or at the very least cause unexpected diffs, by changing order

Reply:
Thank you!! This looks much better and predictable from the PoV of users. Including this. Yeah. In manifest generate test I had to change the order to pass it - https://github.com/istio/istio/blob/02f7faf64f42b041ca8c8c3d188a4a426d0a1a4b/operator/cmd/mesh/manifest-generate_test.go#L185-L187 . With this feedback incorporated I don't have to do that!